test/testaux/utils.py

```diff
@@ -63,32 +58,55 @@ def form_random_constrained_categorical_domain(n_double_param=5, n_int_param=5,
   idx_quantized = idx_shuffled[n_double_param + n_int_param + n_cat_param :]
 
   # Form domain components
-  domain_components = [None] * dim
+  domain_components_unsorted: list[tuple[int, DomainComponent]] = []
   for i in idx_double:
-    bounds = [0, numpy.random.randint(1, 5)]
-    domain_components[i] = {
-      "var_type": DOUBLE_EXPERIMENT_PARAMETER_NAME,
-      "elements": bounds,
-    }
+    bounds = (0, numpy.random.randint(1, 5))
+    domain_components_unsorted.append(
+      (
+        i,
+        {
+          "var_type": "double",
+          "elements": bounds,
+        },
+      )
```

Comment:
what was the problem with the `domain_components[idx]`?

Reply:
I had 2 ways of solving the following problem, maybe you can pick whichever you like better:

We have an array of indices `I` of length `N`. We assume that all these indices are unique and take values `0..N-1`. We want to construct a new array `A` where `A[I[i]] = f(i, I[i]) for i in 0..N-1`. We need to make sure all the elements in `A` are populated (ie. not `None`).

Solution 1:
 1. Initialize `A?[i] = None for i in 0..N-1`
 2. Set `A?[I[i]] = f(i, I[i]) for i in 0..N-1`
 3. Initialize `A[i] = A?[i] for i in 0..N-1`
 4. Assert `A[i] is not None for i in 0..N-1`

Solution 2:
 1. Initialize `M[i] = (I[i], f(i, I[i])) for i in 0..N-1`
 2. Sort `M`
 3. Initialize `A[i] = M[i][1] for i in 0..N-1`
 4. Assert `len(A) == len(I)`

In most cases I went with solution 1

Reply:
> `assert all(dc is not None for dc in domain_components)`

This does not change the type of `domain_components`, it is still a `list[DomainComponent | None]`. 

I have modified the code to:
```python
domain_components_maybe_none: list[DomainComponent | None] = [None] * dim
...
domain_components = [c for c in domain_components_maybe_none if c is not None]
assert len(domain_components) == dim
```

Reply:
great thanks!